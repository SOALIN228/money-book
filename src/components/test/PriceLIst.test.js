import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import PriceList from '../PriceList'
import Ionicon from 'react-ionicons'

const categories = {
  '1': {
    'id': 1,
    'name': '旅行',
    'type': 'outcome',
    'iconName': 'ios-plane'
  },
  '2': {
    'id': 2,
    'name': '理财',
    'type': 'income',
    'iconName': 'logo-yen'
  }
}

const items = [
  {
    'id': 1,
    'title': '旅游',
    'price': 200,
    'date': '2020-05-05',
    'cid': 1
  },
  {
    'id': 2,
    'title': '理财',
    'price': 400,
    'date': '2020-06-05',
    'cid': 2
  }
]

const itemsWithCategory = items.map(item => {
  item.category = categories[item.cid]
  return item
})

const props = {
  items: itemsWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn(),
}

let wrapper
describe('test PriceList component', () => {
  // 每次都会执行的生命周期
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props}/>)
  })
  it('should render the component to match snapshot', () => {
    // 生成快照，每次改变进行比较
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('should render correct price items length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategory.length)
  })
  it('should render correct icon and price for each item', () => {
    const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
    expect(iconList.length).toEqual(3)
    expect(iconList.first().props().icon)
      .toEqual(itemsWithCategory[0].category.iconName)
  })
  it('should trigger the correct function callbacks', () => {
    const firstItem = wrapper.find('.list-group-item').first()
    firstItem.find('a').first().simulate('click', {
      preventDefault: () => {
      }
    })
    expect(props.onModifyItem).toHaveBeenCalledWith(itemsWithCategory[0])
    firstItem.find('a').last().simulate('click', {
      preventDefault: () => {
      }
    })
    expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0])
  })
})

