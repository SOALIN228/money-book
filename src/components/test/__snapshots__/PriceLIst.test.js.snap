// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to match snapshot 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="1"
  >
    <span
      className="col-1 badge badge-primary"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#347eff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      旅游
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      -
      200
      元
    </span>
    <span
      className="col-2"
    >
      2020-05-05
    </span>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="2"
  >
    <span
      className="col-1 badge badge-primary"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="logo-yen"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#347eff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      理财
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      +
      400
      元
    </span>
    <span
      className="col-2"
    >
      2020-06-05
    </span>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
</ul>
`;
